File: LeetCode/416. Partition Equal Subset Sum.py
### KNOWLEDGE IS POWER  ###

class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        summ = sum(nums)
        if summ%2 != 0: 
            return False
        sz = int(summ/2) 
        if sz in nums:  # If Element sum already in array
            return True
        n = len(nums)
        st = set([nums[0]])     # set to generate all subset sums
        for i in range(1, n):
            for j in list(st):
                if nums[i] + j == sz:     # found subset sum
                    return True
                st.add(nums[i]+j)
        return sz in st    

File: LeetCode/416. Partition Equal Subset Sum.java
/***  KNOWLEDGE IS POWER  ***/

class Solution {
    public boolean canPartition(int[] nums) {
        int sum = 0;
        for(int i : nums) 
            sum += i ;
        if(sum%2 != 0)
            return false ;
        int sz = sum/2 ;
        boolean dp[] = new boolean[sz+1] ;
        for(int i = 1 ; i <= sz ; i++)
        {
            dp[i] = false ;
        }
        dp[0] = true ;
        for(int i : nums) {
            for (int j = sz ; j >= i ; j--) 
            {
                if(dp[j - i]) 
                {
                    dp[j] = true ;
                }
            }
        }
        return dp[sz] ;
    }
}
